### ISSUE TITLE
Explore alternatives to Docker Desktop for the engine behind ddev

### ISSUE THREAD

**Issue by rfay** [drud/ddev#3208]
Docker Desktop has stability problems and people are always asking about options.

Explore what the alternatives might be.

Article on Lima: https://medium.com/nttlabs/containerd-and-lima-39e0b64d2a59

/cc @shaal

**Comment by shaal** [comment]
We should also explore if Podman is a viable alternative 
https://github.com/hutger/podman-on-mac

**Comment by rfay** [comment]
Some interesting reads on this topic:

* [An Overview of Docker Desktop Alternatives](https://matt-rickard.com/docker-desktop-alternatives/?utm_source=thenewstack&utm_medium=website&utm_campaign=platform)
* [Docker and the Internet of Entitlement](https://thenewstack.io/this-week-in-programming-docker-and-the-internet-of-entitlement/)

**Comment by greggmarshall** [comment]
Docker Desktop is only for Mac and Windows, right?  Linux doesn't use it?

If so, I wonder if  supporting Windows legacy is a good idea long term (any notion how many Windows legacy installs are still in use?).  Then using Windows with WSL2, switch the documentation to install Docker inside the Linux distribution.

That makes life a bit simpler, leaving only the Mac to deal with (which might be easier to solve than finding a Windows/Mac cross platform option that works well with both).

**Comment by rfay** [comment]
The plan is to move everybody on Windows to WSL2 anyway, regardless of Docker Desktop. So yes, Mac is the issue.

**Comment by CptIceberg** [comment]
> We should also explore if Podman is a viable alternative
> https://github.com/hutger/podman-on-mac

I think this is a great idea since Podman is slowly but surely opening to all platforms possible, and has great usability / performances. (I'm no contributor but a regular user of ddev, thanks for all the great work, you made an incredible tool !)

**Comment by shaal** [comment]
[Goodbye Docker Desktop, Hello Minikube!](https://itnext.io/goodbye-docker-desktop-hello-minikube-3649f2a1c469)

**Comment by deviantintegral** [comment]
I've been watching development of Lima, and it's making good progress. In particular, it now supports forwarding ports < 1024.

I had ddev working inside of lima, and the only real issue was that the `mkcert` setup didn't work because the root cert was added inside the VM. I imagine this could be worked around like how it's done on WSL2 installs.

_however_ I just gave https://github.com/abiosoft/colima a try, and it seems to truly be a drop-in replacement for Docker Desktop. In summary:

1. `brew install colima`
2. I quit docker desktop to ensure the docker socket didn't conflict.
3. `colima start [--cpus 8] [--memory 16]`. This took a few minutes to load, as it was downloading the VM image and docker. I see it as equivalent to downloading Docker Desktop which is a few hundred MB.
4. I went to an existing ddev project, ran `ddev start`, `ddev pull production`, etc. It all just worked!

One note is that I'm using mutagen. As colima uses sshfs for volume mounts, I'm sure that's even slower than `osxfs` and may not be usable at all. Actually, I'm going to rebuild now and try it out.

If anyone can think of areas you'd expect ddev to break by not running Docker Desktop itself, let me know and I'll give it a try.

**Comment by deviantintegral** [comment]
`ddev composer install   0.92s  user 0.40s system 0% cpu 14:25.51 total` - so 14 minutes for `composer install`. At least it works, but I think docker desktop is on the order of 3-5 minutes for a similar project. mutagen still wins by miles!

**Comment by rfay** [comment]
That is completely amazing @deviantintegral !  Could you please write up a step-by-step in https://github.com/drud/ddev-contrib? It could be a start for lots of people, including me.  Thanks so much for working on this!

**Comment by rfay** [comment]
Trying out colima on mac m1, distressed to see it using qemu to run arm64. That's certainly not what we want. But surely can be solved with upstream contributions.

**Comment by mparker17** [comment]
While I understand the majority of people using Ddev probably run macOS, I think there is some merit in keeping Podman-for-macOS in mind, even if it isn't as well-optimized for macOS as colima... hear me out as I try to explain...

***

While I'm not very familiar with GitHub Actions; I know from experience that GitLab CI, Travis, and CircleCI all give you Linux task runners by default. But, because brand-name Docker requires privileged access to run Docker-in-Docker, I cannot use Ddev in CI (even if I am using Ddev locally).

Instead, I have to manually set up a whole LAMP stack in the CI task runner container in order to run Behat tests, Functional and FunctionalJavascript tests, etc. (admittedly, I can do some of this ahead of time / infrequently by using something like Hashicorp Packer to build the LAMP stack, save it to a Docker image, push the Docker image to the project's package registry, and telling the CI task runner to start from that image (i.e.: so all I have to do is install Drupal with the latest code and run the tests when CI runs) but the point is, *I still have to maintain all that LAMP-stack setup myself*).

However, if Ddev used a rootless docker alternative (like Podman), then I could set up Podman on the CI task runner, run `ddev start` and then run all my tests in CI exactly the way I run them on my local... which would make developing and maintaining CI much simpler.

***

I am aware that what I'm advocating for here is actually a particular piece of *software running on the Linux VM to run the containers* (i.e.: brand-name Docker vs Podman); but this GitHub issue is actually for debating the *software to run the Linux VM itself and pass commands through to it* (i.e.: Docker Desktop vs Podman-for-macOS vs colima). 

However, the reason I'm bringing this up here is, if Ddev were to adopt Podman-for-macOS as the *software to run the Linux VM itself and pass commands through to it*, then we'd automatically be adopting Podman as the *software running on the Linux VM to run the containers* - and we would either get the ability to run Ddev in CI for free, or at least reduce the effort needed to do so... and I think this would be a valuable feature for Ddev.

**Comment by rfay** [comment]
* colima runs on linux (I haven't tried it, but they certainly claim it)
* If you can try out/figure out podman with ddev (give it a try) it would be very welcome!
* There is discussion about adding podman as a backend to colima

Related, of course, https://github.com/drud/ddev/issues/2899

Would just rootless docker solve the problem?

**Comment by mparker17** [comment]
> colima runs on linux (I haven't tried it, but they certainly claim it)

Good to know!


> If you can try out/figure out podman with ddev (give it a try) it would be very welcome!

I would be very interested in doing this, but I'm not familiar enough with the inner workings of Ddev to know what to test, or where to start changing things... would you be willing to offer some suggestions / insights / mentorship? :pray: 


> There is discussion about adding podman as a backend to colima

Awesome!


> Related, of course, #2899

I'll check that out, thanks!


> Would just rootless docker solve the problem?

Yes.

I will admit to having a slight preference for Podman, as it is the basis for [Fedora Silverblue](https://silverblue.fedoraproject.org) (an immutable Linux OS I run at home), comes pre-installed on [Fedora IoT](https://getfedora.org/iot/) (which I run on several Raspberry PIs), and has a bunch of open-source drop-in-replacements to other name-brand Docker things I've used in the past; like [`docker-compose`/`podman-compose`](https://github.com/containers/podman-compose) and [Kubernetes' `kubectl`/`podman play kube`](https://docs.podman.io/en/latest/markdown/podman-play-kube.1.html).

(Note that Silverblue and Fedora IoT came out of [OpenShift Container Platform](https://www.redhat.com/en/technologies/cloud-computing/openshift) (formerly [CoreOS Container Linux](https://web.archive.org/web/20170628024331/https://coreos.com/why/))).

**Comment by rfay** [comment]
Well, you don't have to know the internals of ddev to just see if it works. That's what @deviantintegral did with colima. He just started it and tried it out to see what worked. Do it with podman :)

**Comment by rfay** [comment]
And of course I'm happy to help, but just seeing if anything works at all to start would be great.  The things I checked in https://github.com/drud/ddev/issues/3208#issuecomment-965811401 were a starting set. Anything we do will require a lot of effort to make all features work, I'm sure.

**Comment by deviantintegral** [comment]
I don't have an M1 Mac, so I can't comment other than the claimed support of supporting ARM Macs.

> 👎 distressed to see it using qemu to run arm64...

I thinks this has been addressed [in this discussion](https://github.com/lima-vm/lima/discussions/406), which at most is "there's a bug with running amd64 images on arm64" and not a serious design limitation of lima.

> 👎 The initial pull of images is really quite amazingly slow compared to what would happen in any docker environment. Not sure if that might be because it's in docker-in-docker environment?

I've noticed no differences so far (it's maxing out my 50Mbit connection). Perhaps it's an arm specific issue?

> 👎 Doesn't seem to work out of the box with nfs, I'm sure that's solvable.

I'm sure that is, even if it meant layering additional provisioning on top of the lima.yaml file. I'm not sure it's worth it though given the work with mutagen.

> 👎 Mounted filesystem (from macOS) when not using mutagen is read-only, meaning that nothing can be done inside the container (no composer install, no creation of files in sites/default/files)
> 👎 ddev snapshot and ddev pull also can't work because the mounted volume is not writable.

This is documented in colima and if you want a writable filesystem, you can start colima with something like `--mount $HOME/workspace/lullabot:w`. It's sshfs though and very slow (3x slower than `osxfs`!), so again I think a good approach would be to use mutagen.

> 👎 xdebug debugging can't work at this point because host.docker.internal is not defined. I imagine that's solvable.

Good catch. I will try to see if I can find what's recommended for that.

> 👎 elasticsearch fails

That's surprising, any details?

> I think there is some merit in keeping Podman-for-macOS in mind, even if it isn't as well-optimized for macOS as colima

I'm not particularly in the "colima" camp, other than I assumed that podman would likely have more trouble due to the ties between ddev / docker / docker-compose, even it was just calling out to the wrong binaries or not getting the right $PATH. I'd love to try it out, but I likely won't get to that this week. Feel free to run ahead of me!

**Comment by rfay** [comment]
I'm not particularly in any camp either, was just amazed that colima could just come up running.

Thanks for the responses. 

The elasticsearch container just seems to exit (nothing in logs). I was just trying a couple of popular things out so didn't pursue it a lot.

**Comment by deviantintegral** [comment]
The internal host for lima / colima is `host.lima.internal`.

**Comment by tyler36** [comment]
@mparker17 , there is a [Setup DDEV Github action](https://github.com/marketplace/actions/setup-ddev) designed to reuse your DDEV setup in CI.

> This Github action starts ddev with your project's configuration from the directory .ddev.
> 
> The idea is to reuse the same environment that you are maintaining for development anyways for automated acceptance testing, thus saving on maintaining a separate CI-configuration.
> 
> Any additional services that you might have configured will be started and any post-start hooks etc. will be run.

I haven't had a chance to play with it yet but it _might_ suit your need.

**Comment by mparker17** [comment]
@tyler36 that's good to know, but I rarely use GitHub nowadays... my company uses GitLab; and at least 3 client organizations I've worked with recently use internally-hosted GitLab as well.

---

I use Linux as my primary computer; but I have limited control over a mac issued to me by a client organization. I tried installing Podman and Ddev on it, and following [a modified Ddev quickstart scenario](https://ddev.readthedocs.io/en/stable/users/cli-usage/#quickstart-guides)

```
$ brew install podman
$ brew install drud/ddev/ddev
$ pip3 install --user podman-compose
$ echo 'export PATH="$PATH:$HOME/Library/Python/3.8/bin"' | tee -a .zshrc
$ # Restarted my terminal

$ podman machine init
$ # This took a few minutes to download the VM image, which is Fedora CoreOS Stable. I forgot to time it; sorry.
$ # Initializing a machine after it has been downloaded takes about 22 seconds

$ podman machine start
INFO[0000] waiting for clients...
INFO[0000] listening tcp://0.0.0.0:7777
INFO[0000] new connection from  to /var/folders/yx/<random>/T/podman/qemu_podman-machine-default.sock
Waiting for VM...
Machine "podman-machine-default" started successfully
$ # podman machine start took about 8 seconds

$ cd $HOME/Development
$ mkdir ddev && cd ddev
$ mkdir my-new-project && cd my-new-project
$ ddev config --project-name=my-new-project --create-docroot --docroot=web --project-type=drupal8

$ ddev start
Could not connect to docker. Please ensure Docker is installed and running.

$ alias docker=podman
$ alias docker-compose=podman-compose
$ ddev start
Could not connect to docker. Please ensure Docker is installed and running.
```

My guess is that, because Podman does not try to pass itself off as brand-name Docker (i.e.: `which docker` and `which docker-compose` return "Command not found" before I alias), when `cmd/ddev/cmd/root.go::PersistentPreRun()` calls `pkg/dockerutil/dockerutils.go::CheckDockerVersion()`, it cannot detect a version.

---

Some other notes:

1. The Linux guest VM is the latest [Fedora CoreOS from the testing stream](https://getfedora.org/en/coreos?stream=testing) - in this case 20211029.2.0. Since I have an x86_64 mac, it used that architecture.
2. More information about how Podman-for-macOS works is in [this article, *How Podman runs on Macs and other container FAQs*](https://www.redhat.com/sysadmin/podman-mac-machine-architecture). Briefly, it uses [QEMU with HVF acceleration](https://wiki.qemu.org/Features/HVF) - HVF uses the macOS [Hypervisor.framework](https://developer.apple.com/documentation/hypervisor); [Ignition](https://coreos.github.io/ignition/) to configure the VM; and [gvisor-tap-vsock](https://github.com/containers/gvisor-tap-vsock) to forward ports, 
3. [Podman does support Apple Silicon/aarm64](https://podman.io/blogs/2021/10/04/m1macs.html) - and Fedora CoreOS is available for the aarch64 (i.e.: ARM64/ARMv8-A/Apple M1) architecture as well

---

Trying to address specific things brought up by @rfay and @deviantintegral...

> Support for forwarding ports < 1024

Podman docs are unclear about how this should work.

On Linux, I can `sudo podman run -dt -p 80:80/tcp --net=host nginx` but this fails on macOS with the following error...

```
Port mappings have been discarded as one of the Host, Container, Pod and None network modes are in use
```

... and its unclear how to resolve this.

I've run out of time but I'd check https://github.com/containers/gvisor-tap-vsock next, to see if there's any more info there.


> qemu to run arm64

While I wasn't able to test it on arm64; and I'm not sure I fully understand the question:

1. it Podman downloaded the guest Linux OS for the same architecture as the host machine;
2. I didn't put a lot of work into seeing if I could get it to emulate the arm64 architecture on my x86_64 host machine
3. I wasn't able to test it on arm64
4. Podman uses [QEMU with HVF acceleration](https://wiki.qemu.org/Features/HVF); HVF uses the official macOS [Hypervisor.framework](https://developer.apple.com/documentation/hypervisor)


> pulling images

I haven't noticed any issues... seems pretty fast.


> using mutagen
> using nfs
> read-only when not using mutagen

There's some discussion of using podman with mutagen, nfs, etc. in containers/podman#8016

Mutagen is https://mutagen.io right? I've no experience using this.

[Apparently](https://www.redhat.com/sysadmin/rootless-podman-nfs), rootless and rootfull podman volume sharing works great with NFS; but out-of-the-box, rootless podman doesn't work well when you store the container images on an NFS mount point - although it is possible to work around that.


> ddev snapshot and ddev pull
> xdebug
> inter-project communication
> solr
> elasticsearch
> using https/mkcert 
> ddev auth ssh

Since I wasn't able to get Ddev working locally, I wasn't able to check these things in a realistic way.

---

Hopefully that helps?

**Comment by mbomb007** [comment]
Question: What is the planned/approximate timeline for when Docker Desktop would no longer be needed on Windows?

**Comment by rfay** [comment]
Docker Desktop is already not needed on Windows. WSL2 is the future of DDEV support on Windows, and although the current recipe uses Docker Desktop, lots of people just install (Linux) docker inside WSL2 and it works great there. I imagine we'll switch to recommending that in the future.

**Comment by tyler36** [comment]
Are there instructions for "just install (Linux) docker inside WSL2".
I tried adding it via brew

```shell
$brew install docker

$ ddev start
Could not connect to docker. Please ensure Docker is installed and running.
```

If DDEV is heading towards a WSL2-prefered solution, we should be offering instructions now so they can be battle-tested and refined.

**Comment by rfay** [comment]
Use the Linux installation instructions, https://ddev.readthedocs.io/en/latest/users/docker_installation/#linux-installation-docker-and-docker-compose

**Comment by tyler36** [comment]
Thank you. Seems obvious after I read it.

**Comment by greggmarshall** [comment]
@tyler36 I did the docker in Linux on yet another contract supplied Windows 10 machine last week and it and DDEV installed without any issues.  One recommendation I have is to set the memory for your WSL2 environment in .wslconfig to no more than half your computer's memory.  I had a previous computer where I didn't do that and WSL2 eventually took so much memory it effectively locked up the machine.

**Comment by mbomb007** [comment]
> Use the Linux installation instructions, https://ddev.readthedocs.io/en/latest/users/docker_installation/#linux-installation-docker-and-docker-compose

I tried doing this with Ubuntu-20.04 WSL2. However, when I get to the point where I'm supposed to verify that it works, the docker service isn't running. The instructions don't say how to start the docker service on WSL2.

```
docker context ls
NAME        DESCRIPTION                               DOCKER ENDPOINT               KUBERNETES ENDPOINT   ORCHESTRATOR
default *   Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                         swarm
```

`env | grep DOCKER_HOST` returns nothing.
```
service docker start
docker: unrecognized service
```

There doesn't appear to be `docker.sock` in `/var/run`.

So I really have no idea where to go from here.

**Comment by rfay** [comment]
This isn't the place for a discussion about docker installation on Linux, but see https://docs.docker.com/engine/install/ubuntu/, I imagine you'll need to `systemctl start docker` and enable it I think. 

Happy to help on Discord or one of the other support channels, but note that I just found out today that wsl2 with docker installed inside it can't do xdebug on v1.18.1 (although it can on all earlier versions). So you'd want to use the `install_ddev.sh v1.18.0` with it.

**Comment by mbomb007** [comment]
The [install instructions](https://ddev.readthedocs.io/en/stable/users/docker_installation/#linux-installation-docker-and-docker-compose) said

> Ubuntu 20.04+ has recent enough versions that you can `sudo apt-get update && sudo apt-get install docker.io docker-compose`

However, this is not a detailed guide, and it's not enough to run only that command. So basically, you still need to go through the Ubuntu-specific install instructions, even for 20.04. Thanks.

**Comment by rfay** [comment]
This is the wrong issue for this conversation. Please open a new one, do a docs PR, or use one of the many other support venues, https://ddev.readthedocs.io/en/latest/#support-and-user-contributed-documentation - I'm happy to help, but not in an issue that is devoted to something else entirely.

**Comment by rfay** [comment]
* https://github.com/drud/ddev/issues/3419 is a starter for docker-inside-wsl2 docs.

**Comment by deviantintegral** [comment]
> The internal host for lima / colima is `host.lima.internal`.

I filed an upstream feature request for this at https://github.com/lima-vm/lima/issues/456. I'd like to give the maintainers a few days to respond, and I will see if someone on our team is interested in writing docs for the contrib section.

**Comment by rfay** [comment]
Some questions that people will want answered for any candidate:
- [ ] Can I use it with PhpStorm?
- [ ] Can I use it with vscode?
- [ ] Can I debug using xdebug?
- [ ] Do web browsers run inside the environment (like lima) or can the run on the host computer?
- [ ]

**Comment by deviantintegral** [comment]
>  Do web browsers run inside the environment (like lima)

Not sure what you mean by this. I can access ddev services from my macOS browser without issue. Of course, we can _also_ run containers with browsers for functional tests. Were you doing something like that, but with VNC or X11 forwarding?

**Comment by rfay** [comment]
In the WSL2 world, which is a bit similar to some of what we're doing here, there are lots of ways to do things. You can run browser or PhpStorm inside WSL2 or run them on the Windows side. I agree that right now colima at least seems like nearly a drop-in replacement for Docker Desktop, just using the host-side tools.

**Comment by deviantintegral** [comment]
I found that running with a writable mount was very slow for composer operations. Did it still seem fast enough for some other framework?

With PHPStorm's new support for backend remotes over SSH, it seems like it would also be possible to use that. I can see `colima ssh` is _actually_ ssh (and not some other socket or process), so it should be possible.

**Comment by rfay** [comment]
I spent just a little time with podman this morning. 
* Like lima and Docker Desktop, podman runs a linux vm and all the real work is done there.
* podman still provides no solution for filesystem mounting, so seems unlikely to be a match for ddev right now.
* Unlike lima/colima, the docker api does not seem to be supported, so places where ddev uses the docker api don't seem to work
* But it works fine for running containers, even arm64 containers.
* colima has a [podman runtime PR](https://github.com/abiosoft/colima/pull/27), not sure how they'll do that with no filesystem mounting.

**Comment by rfay** [comment]
Just visiting with minikube, as it's often recommended. 

The reality is that minikube just runs on top of another container provider, so 
* on amd64 it has the drivers "virtualbox, vmwarefusion, kvm2, vmware, none, docker, podman, ssh"
* on arm64 is has the drivers "docker, podman (experimental), ssh"

So it doesn't seem to have value the way ddev is currently run (with compose). It obviously might if ddev switched from docker-compose to kubernetes in the future, which is certainly a possibility.

**Comment by deviantintegral** [comment]
> colima has a podman runtime PR, not sure how they'll do that with no filesystem mounting.

That PR is to support the "linux" podman runtime, and not to use the podman VM wrapper. So, since filesystem mounting is handled by colima and lima, it shouldn't be any different than with docker.

**Comment by rfay** [comment]
Updated the bit about minikube in https://github.com/drud/ddev/issues/3208#issuecomment-990032541

**Comment by deviantintegral** [comment]
I'm surprised minikube appears to work as well! I misunderstood that it could work for docker composer, and not just kubernetes. What did you run into with on arm?

I spent some time trying PHPStorm's remoting functionality with colima. It worked reasonably well, and I'd say it's a similar workflow to using X11 forwarding on WSL2. Performance was fine, but one gap is that for Intel processors, you don't want to expose all cores to the VM (typically VMs can't tell if a core is hyperthreaded or not, messing up thread scheduling), so indexing was a bit slower. However, I was able to break things pretty quickly with a "find usages" call that seemed to hang the remote. It's probably worth looking into and documenting officially when it hits a "stable" declaration from JetBrains.

**Comment by rfay** [comment]
Minikube/hyperkit on arm64:

```
$ minikube start --driver=hyperkit
😄  minikube v1.24.0 on Darwin 12.0.1 (arm64)
✨  Using the hyperkit driver based on user configuration

❌  Exiting due to DRV_UNSUPPORTED_OS: The driver 'hyperkit' is not supported on darwin/arm64
```

**Comment by rfay** [comment]
Using PhpStorm's remoting into colima is essentially running *linux* ddev inside colima, right? Which is very much like WSL2, or vscode with its remoting setup.

**Comment by deviantintegral** [comment]
Yes, exactly. Of note is that ddev defaulted to http, not https URLs, and can't interact with the cert store on macOS. We'd need instructions on how to do that.

**Comment by rfay** [comment]
hyperkit: https://github.com/moby/hyperkit/issues/310#issuecomment-880492534

I imagine the instructions for trusting mkcert will be something like we have to [do with $CAROOT](https://ddev.readthedocs.io/en/latest/#installation-or-upgrade-windows-wsl2) in WSL2.  Also the upstream instructions for the same basic thing, https://github.com/FiloSottile/mkcert#installing-the-ca-on-other-systems

**Comment by rfay** [comment]
I spent some time with podman 3.0.1 on Linux (Debian 11) and found that while it is definitely a docker replacement, it is absolutely not a drop-in replacement. I used $DOCKER_HOST to point to the correct endpoint, and was able to use both the podman and docker client tools, and was able to get ddev to connect and start up. There were a few hiccups along the way, but the bottom line was that podman has extensive differences and problems, and I don't know how to get past them, even on its home ground, linux.
